File: game/src/main/java/com/bship/games/domains/Board.java
package com.bship.games.domains;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;

import java.util.List;
import java.util.Objects;

import static com.bship.games.util.Util.addTo;

@JsonDeserialize(builder = Board.Builder.class)
public class Board {

    @JsonIgnore
    private Long gameId;
    private Long id;

    private List<Piece> pieces;
    private List<Piece> opponentPieces;
    private List<Move> moves;
    private List<Move> opponentMoves;
    private final boolean winner;

    private Board(Builder builder) {
        id = builder.id;
        gameId = builder.gameId;
        pieces = builder.pieces;
        opponentPieces = builder.opponentPieces;
        moves = builder.moves;
        opponentMoves = builder.opponentMoves;
        winner = builder.winner;
    }

    public Long getGameId() {
        return gameId;
    }

    public Long getId() {
        return id;
    }

    public List<Piece> getPieces() {
        return pieces;
    }

    public List<Piece> getOpponentPieces() {
        return opponentPieces;
    }

    public List<Move> getMoves() {
        return moves;
    }

    public List<Move> getOpponentMoves() {
        return opponentMoves;
    }

    public boolean isWinner() {
        return winner;
    }

    public Builder copy() {
        return builder()
                .withId(getId())
                .withGameId(getGameId())
                .withPieces(getPieces())
                .withOpponentPieces(getOpponentPieces())
                .withMoves(getMoves())
                .withOpponentMoves(getOpponentMoves())
                .withWinner(isWinner());

    }

    public static Builder builder() {
        return new Builder();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Board that = (Board) o;

        return Objects.equals(this.gameId, that.gameId) &&
                Objects.equals(this.id, that.id) &&
                Objects.equals(this.moves, that.moves) &&
                Objects.equals(this.opponentMoves, that.opponentMoves) &&
                Objects.equals(this.opponentPieces, that.opponentPieces) &&
                Objects.equals(this.pieces, that.pieces) &&
                Objects.equals(this.winner, that.winner);
    }

    @Override
    public int hashCode() {
        return Objects.hash(gameId, id, moves, opponentMoves, opponentPieces, pieces,
                winner);
    }

    @JsonPOJOBuilder
    public static final class Builder {
        private Long id;
        private Long gameId;
        private List<Piece> pieces;
        private List<Piece> opponentPieces;
        private List<Move> moves;
        private List<Move> opponentMoves;
        private boolean winner;

        public Builder withId(Long id) {
            this.id = id;
            return this;
        }

        public Builder withGameId(Long gameId) {
            this.gameId = gameId;
            return this;
        }

        public Builder addPiece(Piece piece) {
            pieces = addTo(pieces, piece);
            return this;
        }

        public Builder withPieces(List<Piece> pieces) {
            this.pieces = pieces;
            return this;
        }

        public Builder withOpponentPieces(List<Piece> opponentPieces) {
            this.opponentPieces = opponentPieces;
            return this;
        }

        public Builder addOpponentPieces(Piece piece) {
            opponentPieces = addTo(opponentPieces, piece);
            return this;
        }

        public Builder addMove(Move move) {
            moves = addTo(moves, move);
            return this;
        }

        public Builder withMoves(List<Move> moves) {
            this.moves = moves;
            return this;
        }

        public Builder withOpponentMoves(List<Move> opponentMoves) {
            this.opponentMoves = opponentMoves;
            return this;
        }

        public Builder addOpponentMove(Move move) {
            opponentMoves = addTo(opponentMoves, move);
            return this;
        }

        public Builder withWinner(boolean winner) {
            this.winner = winner;
            return this;
        }

        public Board build() {
            return new Board(this);
        }
    }

    @Override
    public String toString() {
        return "{" +
                "\"gameId\":" + getGameId() +
                ", \"id\":" + getId() +
                ", \"ships\":" + getPieces() +
                ", \"opponentPieces\":" + getOpponentPieces() +
                ", \"moves\":" + getMoves() +
                ", \"opponentMoves\":" + getOpponentMoves() +
                ", \"winner\":" + isWinner() +
                '}';
    }
}

File: game/src/main/java/com/bship/games/endpoints/cabinet/entity/Board.kt
package com.bship.games.endpoints.cabinet.entity

import com.bship.games.Configuration.BoardDeserializer
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import java.util.*
import java.util.Collections.emptyList

@JsonDeserialize(using = BoardDeserializer::class)
class Board private constructor(val id: Long,
                                val gameId: Long,
                                val pieces: List<Piece>,
                                val opponentPieces: List<Piece>,
                                val moves: List<Move>,
                                val opponentMoves: List<Move>,
                                val winner: Boolean) {

    private constructor(builder: Builder) : this(
            id = builder.id.orElse(0),
            gameId = builder.gameId.orElse(0),
            pieces = builder.pieces.orElse(emptyList()),
            opponentPieces = builder.opponentPieces.orElse(emptyList()),
            moves = builder.moves.orElse(emptyList()),
            opponentMoves = builder.opponentMoves.orElse(emptyList()),
            winner = builder.winner.orElse(false)
    )

    fun copy(init: Builder.() -> Unit): Board {
        return Builder(this).apply(init).build()
    }

    class Builder() {
        constructor(board: Board) : this() {
            withId { board.id }
            withGameId { board.gameId }
            withWinner { board.winner }
            withMoves { board.moves }
            withPieces { board.pieces }
            withOpponentMoves { board.opponentMoves }
            withOpponentPieces { board.opponentPieces }
        }

        internal var id: Optional<Long> = Optional.empty(); private set
        internal var gameId: Optional<Long> = Optional.empty(); private set
        internal var winner: Optional<Boolean> = Optional.empty(); private set
        internal var moves: Optional<List<Move>> = Optional.empty(); private set
        internal var pieces: Optional<List<Piece>> = Optional.empty(); private set
        internal var opponentMoves: Optional<List<Move>> = Optional.empty(); private set
        internal var opponentPieces: Optional<List<Piece>> = Optional.empty(); private set

        fun withId(init: Builder.() -> Long?) = apply {
            id = Optional.ofNullable(init())
        }

        fun withGameId(init: Builder.() -> Long?) = apply {
            gameId = Optional.ofNullable(init())
        }

        fun withPieces(init: Builder.() -> List<Piece>?) = apply {
            pieces = Optional.ofNullable(init())
        }

        fun withOpponentPieces(init: Builder.() -> List<Piece>?) = apply {
            opponentPieces = Optional.ofNullable(init())
        }

        fun withMoves(init: Builder.() -> List<Move>?) = apply {
            moves = Optional.ofNullable(init())
        }

        fun withOpponentMoves(init: Builder.() -> List<Move>?) = apply {
            opponentMoves = Optional.ofNullable(init())
        }

        fun withWinner(init: Builder.() -> Boolean?) = apply {
            winner = Optional.ofNullable(init())
        }

        fun build() = Board(this)
    }


    companion object {
        inline fun build(init: Builder.() -> Unit): Board =
                Builder().apply(init).build()
    }

    override fun toString(): String {
        return "{" +
                "\"gameId\":" + gameId +
                ", \"id\":" + id +
                ", \"ships\":" + pieces +
                ", \"opponentPieces\":" + opponentPieces +
                ", \"moves\":" + moves +
                ", \"opponentMoves\":" + opponentMoves +
                ", \"winner\":" + winner +
                '}'
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Board

        if (id != other.id) return false
        if (gameId != other.gameId) return false
        if (pieces != other.pieces) return false
        if (opponentPieces != other.opponentPieces) return false
        if (moves != other.moves) return false
        if (opponentMoves != other.opponentMoves) return false
        if (winner != other.winner) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + gameId.hashCode()
        result = 31 * result + pieces.hashCode()
        result = 31 * result + opponentPieces.hashCode()
        result = 31 * result + moves.hashCode()
        result = 31 * result + opponentMoves.hashCode()
        result = 31 * result + winner.hashCode()
        return result
    }
}

